cpp/cmake_modules/DefineOptions.cmake

```diff
@@ -0,0 +1,333 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+macro(define_option name description default)
+  option(${name} ${description} ${default})
+  list(APPEND ARROW_ALL_OPTION_NAMES ${name})
+  set("${name}_OPTION_DESCRIPTION" ${description})
+  set("${name}_OPTION_DEFAULT" ${default})
+endmacro()
+
+# Top level cmake dir
+if("${CMAKE_SOURCE_DIR}" STREQUAL "${CMAKE_CURRENT_SOURCE_DIR}")
+  set(ARROW_CXXFLAGS "" CACHE STRING "Compiler flags to append when compiling Arrow")
```

Comment:
Can we show no ON/OFF build options such as this?

Reply:
@kou Yes, I can include `ARROW_CXXFLAGS` in the output. Are there any other variables you'd like to see?

Reply:
I'll add `ARROW_CXXFLAGS`, `ARROW_TEST_LINKAGE`, `ARROW_DEPENDENCY_SOURCE`, `ARROW_WITH_ZSTD`, `ARROW_GANDIVA_PC_CXX_FLAGS`, and [these msvc options](https://github.com/apache/arrow/pull/4305/files#diff-b048bf4c1679dce1028fd897a7c43b93R200)

Reply:
Thanks!